ContentFilteredTopic Design Proposal. (#282)
ros2/design: articles/content_filtering.md

- User frontend (rclcpp/rclpy)
  User API will be added to manage filtering configuration, so that user application can set its own filter_expression and expression_parameters. Also using AsyncParametersClient::on_parameter_event, user can take care of the parameter event with user callback only for filtered parameter events.
  **[T.B.D] Filtering should be compatible for user even if rmw_implementation does not support ContentFilteredTopic, this can be done with using ParameterEventsFilter. But the problem here is ParameterEventsFilter only filters parameter name and event type such as NEW, DELETED, CHANGED. there is a huge gap of the flexibility that dds filtering expression and parameter expression possesses. This is what we need to discuss.**

<img src="../img/content_filter/parameter_event_sequence_with_filter.png">

### Interfaces

#### rcl

- add API `is_cft_supported` to allow frontend(rclcpp/rclpy) can know CFT is supported or not.
  e.g) `rcl_subscription_is_cft_supported` to access `is_cft_supported`.
- add `rcl_create_cft_subscription` to call `rmw_create_cft_subscription`.

> - add `rcl_cft_set/get_expression_parameters` to call `rmw_cft_set/get_expression_parameters`.

Reviewer: Having a `rcl_is_cft_supported` function sounds like a good idea, but I don't get why we need all the others.

A cft expression in the subscription options sounds like a better idea than a special function to create a cft_subscription.
The `cft` expression can be a custom object, and we can add a proper API to configure it.

Author: user application is likely to change filtering expression and parameter expressions at runtime, so i think that we need to provide methods to do that dynamically.

Reviewer: > user application is likely to change filtering expression and parameter expressions at runtime, so i think that we need to provide methods to do that dynamically.

Do you mean changing the expression in an already created subscription?
Or dynamically creating an expression?

In the first case, having a special object for the `cft` subscription makes sense.
In the second, only having a object for the expression makes sense.

Author: I am not sure if i follow you precisely,

> Do you mean changing the expression in an already created subscription?

I say yes, this is needed especially for action topics. filtering expression and expression parameters for goal id need to be adjusted at runtime.

Reviewer: Oh yeah, we need to be able to modify the filtering expression of an already created cft subscription (or at least, the expression parameters).

#### rmw

- add new member `is_cft_supported` into `rmw_subscription_t` which indicates if CTF is supported by rmw_implementation or not.
  e.g) see `can_loan_masseges` member as reference.
  support status will be set when parent subscription is created. then support status can be used to notify the upper layer if CFT is supported or not.

| Implementation | `is_cft_supported` |
| :--: | :--: |
| rmw_connext | true |
| rmw_fastrtps | false |
| rmw_cyclonedds | false |

- new `DDSContentFilteredTopic` object needs to be managed in implementation.
- parent topic information should be cached so that we can create CFT based on that parent topic.
- `rmw_create_cft_subscription` interface.
  - input: rmw_subscription_t, rmw_node_t, parent topic information.
  - output: rmw_subscription_t (with new data reader)

> - `rmw_cft_set/get_expression_parameters` interface.

Reviewer: similar comments to `rcl` API

>   - input: `DDSContentFilteredTopic`, (string sequence) expression parameters.

Reviewer: types in `rmw` shouldn't be a DDS type.

Author: correct, that should be `rmw_subscriotion`, that will be fixed.

- Filtering Expression and Parameters
  User can specify node name and parameter name to filter the parameter evetns.
  - [T.B.D] This needs to be discussed more, if user should be able to filter arbitrary parameter or not. if to support any expression and parameter, it means that `rcl` needs to have DDS standard expression and parameter filtering function. 

<img src="../img/content_filter/parameter_event_sequence_with_filter.png">

### Interfaces

#### rcl

- add API `is_cft_supported` to allow frontend(rclcpp/rclpy) can know CFT is supported or not.
  e.g) `rcl_subscription_is_cft_supported` to access `is_cft_supported`.

> - add `rcl_create_cft_subscription` to call `rmw_create_cft_subscription`.

Reviewer 2: @fujitatomoya 

Based on your design, I have a question about the function `rmw_create_cft_subscription`.
If using this function, a subscriber can dynamically use `rmw_create_cft_subscription` to set the [topic_reader_](https://github.com/ros2/rmw_connext/blob/25721877d133a6ea09c1f3a864ed892d288fa918/rmw_connext_cpp/src/rmw_subscription.cpp#L224-L225) inside it at any time, which means we need to add an extra lock to protect the `subscriber_info[topic_reader_, read_condition_]` in `rmw_create_cft_subscription` and some other locations like [rmw_connext-wait](https://github.com/ros2/rmw_connext/blob/25721877d133a6ea09c1f3a864ed892d288fa918/rmw_connext_shared_cpp/include/rmw_connext_shared_cpp/wait.hpp#L224-L230).

I was thinking that using `rmw_create_subscription` can create normal topic or content filtered topic based on rmw_subscription_options_t that adding `const char * filter_expression` and `const rcutils_string_array_t * expression_parameters` if CFT is supported. Is it necessary to separate this operation into two actions?

Author: I think that is fair and reasonable just to extend the interface to support cft subscription. I will update the design as you suggested.

### Parameter Event Topic

Each node will publish (this is configurable via `NodeOptions`) and subscribe parameter events topic with `AsyncParametersClient` or `ParameterEventHandler`.
This topic is to support monitoring parameters for any parameter changes.
It is expected that client libraries will implement the ability to register a callback for specific parameter changes using this topic.
And there is an internal subscription for each node for `/parameter_events` topic to handle the TimeSource via "use_sim_time" parameter.
`rclcpp::ParameterEventsFilter` is used to filter the parameter with specific name "use_sim_time".
As user interface, `AsyncParametersClient::on_parameter_event` is provided to handle parameter events.

<img src="../img/content_filter/parameter_event_overview.png" width="500">

The problem can be broken into the following two parts,

- Network resource consumption on `/parameter_events` topic
  Since publisher does not know who needs to receive which messages, all of the parameter activity will be published and subscribed.

>   That said everyone publishes and subscribes all of the events within this giant topic `/parameter_events`.

Reviewer: ```suggestion
  That said everyone publishes and subscribes all of the events in this high traffic `/parameter_events` topic.
```

  This leads to a lot of unnecessary message transmission over the network.
- Subscriber needs to filter unnecessary (not interested) messages
  Subscriber does not know if the message is something it needs or not without checking the contents of the message via user callback.
  This filtering needs to be done every single time the message comes in.
  In expectation, it is likely that parameter event callback is interested on specific parameters only, so that there will be a lot of unnecessary message receives and filtering process needs to be done.

As described above, we could imagine that if the number of node are 10, 20 and 100 and so are the parameters for each nodes.
This will be a huge burden and pain for entire system, passing unnecessary messages via limited bandwidth network with edge devices and filtering messages to drop is not something user application wants to do.

Besides, this problem can be applied to user defined topics.

### Action Topics

> Each action server will provide two topics named `feedback` and `status` as followings,

Reviewer: ```suggestion
Each action server provides two topics named `feedback` and `status` as followings,

```

<img src="../img/content_filter/action_architecture_overview.png" width="600">

> `feedback` and `status` are topics mapped with action name and published by action server, and action clients subscribe those topics to get feedback and status.

Reviewer: ```suggestion
`feedback` and `status` are topics mapped with action name and published by action server, and action clients subscribe to those topics to get feedback and status.

```

  This is also said that ContentFilteredTopicDataReader does need to do filtering always just in case ContentFilteredTopicDataWriter ends up sending w/o filtering.
  <img src="../img/content_filter/contents_filter_overview.png" width="600">
- The filtering is done by means of evaluating a logical expression that involves the values some of the data-fields in the sample.
  The logical expression is derived from the filter_expression and expression_parameters arguments.
- Each ContentFilteredTopic has a single filter_expression and a sequence of expression_parameters.
  ```
  [parameter use cases]
  "node = %0 OR node = %1" spinal_node attention_node : subscribe all parameter events from spinal and attention nodes
  "node = %0 OR name = %1" audio_node attention_level : subscribe all parameter events from audio_node and parameter named attention_level of all nodes.
  "(node = %0 AND name = %1) OR (node = %2 AND name = %3)" eye_node brightness ear_node loudness : subscribe brightness parameter of eye_node and loudness parameter of ear_node only.
  [action use case]
  "uuid = %0 OR uuid = %1" DEADBEEF CAFEFEED : subscribe only interest goal ids which client handlers possess.
  ```

> - A single filter_expression and a sequence of expression_parameters can be modified dynamically at runtime.

Reviewer: I think it would be good to have a full specification of what's a valid filter_expression, even if it's a link to the specification used by DDS.

Author: that is actually good idea.

- Subscriber is responsible to create ContentFilteredTopic.
- ContentFilteredTopicDataWriter does filtering for each ContentFilteredTopicDataReader endpoint.
  Basically it is likely that ContentFilteredTopicDataWriter does filtering on writer side, so that it can provide the improvement and efficiency.
  But it is dependent on imnplementation if always ContentFilteredTopicDataWriter does filtering.
  This is because if there is a lot of ContentFilteredTopicDataReader with complicated filtering expression and parameter, it would be probably more efficient just to send the messages to all ContentFilteredTopicDataReader.
  And ContentFilteredTopicDataReader can filter out the message if not necessary based on filtering expression and parameter.
  This is also said that ContentFilteredTopicDataReader does need to do filtering always just in case ContentFilteredTopicDataWriter ends up sending w/o filitering.
  <img src="../img/content_filter/contents_filter_overview.png" width="600">
- The filtering is done by means of evaluating a logical expression that involves the values some of the data-fields in the sample.

>   The logical expression is derived from the filter_expression and expression_parameters arguments.

Reviewer 3: When using parameters with multiple DDS implementations note that some differences have been encountered.  See https://issues.omg.org/issues/spec/DDS/1.4#issue-47484

Author: appreciate for the heads-up. i will keep that in the note section.

> - Multiple filter_expression and expression_parameters can be supported on single ContentFilteredTopic.

Reviewer 3: Each ContentFilteredTopic has a single filter_expression and a sequence of expression_parameters

Author: thanks, i borrowed your sentence here.

| flags | cpu | memory | tx (Network load) | improvement |
| --- | --- | --- | --- | --- |
| 1/10 'yes' | 0.40% | 0.2% | 11.49 KByte/s | 90% |
| 2/10 'yes' | 0.46% | 0.2% | 22.9 KByte/s | 80% |
| 3/10 'yes' | 0.48% | 0.2% | 34.28 KByte/s | 70% |
| 4/10 'yes' | 0.55% | 0.2% | 45.69 KByte/s | 60% |
| 5/10 'yes' | 0.60% | 0.2% | 57.09 KByte/s | 50% |
| 6/10 'yes' | 0.69% | 0.2% | 68.49 KByte/s | 40% |
| 7/10 'yes' | 0.73% | 0.2% | 79.89 KByte/s | 30% |
| 8/10 'yes' | 0.79% | 0.2% | 91.29 KByte/s | 20% |
| 9/10 'yes' | 0.84% | 0.2% | 102.69 KByte/s | 10% |
| 10/10 'yes' | 0.85% | 0.2% | 114.09 KByte/s | 0% |
| no filter | 0.85% | 0.2% | 114.05 KByte/s | Baseline |

![improvement_cpu](../img/content_filter/improvement_cpu.png)
![improvement_memory](../img/content_filter/improvement_memory.png)
![improvement_tx](../img/content_filter/improvement_tx.png)

**The bigger data it handles, the more improvement we can have.**

### Specification

#### DDS 

the following create/delete API's are defined,

- create_contentfilteredtopic()
- delete_contentfilteredtopic()

**According to the specification, filter_expression can be only initialized at constructor and it is read_only.

> Probably it is not possible to change the filter_expression at runtime based on the specification but [RTI v5.3.0](https://community.rti.com/static/documentation/connext-dds/5.3.0/doc/api/connext_dds/api_cpp2/classdds_1_1topic_1_1ContentFilteredTopic.html#a76310bf0b7123dd89afbacf43dbabf4a) can support this requirement.

Reviewer 3: filter_expression can be changed at runtime
If the spec is unclear you could consider filing an issue https://issues.omg.org/issues/create-new-issue?specification=DDS%231.4

Author: @mitza-oci thanks for the info. this was kinda confirmation since i was not clear on this.
